File: hooks/scripts/pre/http-proxy.mjs
#!/usr/bin/env node
import { execFileSync } from 'node:child_process';
import { deny, passthrough, readStdin } from '../_lib.mjs';

const TIMEOUT_SECONDS = '5';

const url = (process.env.CCA_HTTP_HOOK_URL ?? '').trim();
if (!url) passthrough();

const data = readStdin();
if (!data) passthrough();

const token = (process.env.CCA_HTTP_HOOK_TOKEN ?? '').trim();
const failClosed = (process.env.CCA_HTTP_HOOK_FAIL_CLOSED ?? '').trim() === '1';

const body = JSON.stringify(data);

const curlArgs = [
  '-s',
  '-X', 'POST',
  '-H', 'Content-Type: application/json',
  '--max-time', TIMEOUT_SECONDS,
  '--write-out', '\n%{http_code}',
  '--data', body,
];
if (token) {
  curlArgs.push('-H', `Authorization: Bearer ${token}`);
}
curlArgs.push(url);

try {
  const output = execFileSync('curl', curlArgs, { encoding: 'utf8' });
  const lines = output.trimEnd().split('\n');
  const statusCode = parseInt(lines[lines.length - 1], 10);
  const respBody = lines.slice(0, -1).join('\n').trim();

  if (statusCode === 204) passthrough();

  if (!respBody) passthrough();

  let result;
  try {
    result = JSON.parse(respBody);
  } catch {
    passthrough();
  }

  process.stdout.write(`${JSON.stringify(result)}\n`);
  process.exit(0);
} catch {
  if (failClosed) {
    deny('[http-hook] Enterprise DLP server unreachable and CCA_HTTP_HOOK_FAIL_CLOSED=1. Blocking action.');
  }
  passthrough();
}

File: hooks/scripts/pre/http-proxy.py
#!/usr/bin/env python3
import json
import os
import sys
import urllib.request
import urllib.error

sys.path.insert(0, os.path.dirname(__file__))
from _lib import read_stdin, deny, passthrough

TIMEOUT = 5


def main() -> None:
    url = os.environ.get("CCA_HTTP_HOOK_URL", "").strip()
    if not url:
        passthrough()

    data = read_stdin()
    if not data:
        passthrough()

    token = os.environ.get("CCA_HTTP_HOOK_TOKEN", "").strip()
    fail_closed = os.environ.get("CCA_HTTP_HOOK_FAIL_CLOSED", "").strip() == "1"

    body = json.dumps(data).encode("utf-8")
    headers = {"Content-Type": "application/json"}
    if token:
        headers["Authorization"] = f"Bearer {token}"

    req = urllib.request.Request(url, data=body, headers=headers, method="POST")

    try:
        with urllib.request.urlopen(req, timeout=TIMEOUT) as resp:
            if resp.status == 204:
                passthrough()
            resp_body = resp.read().decode("utf-8", errors="replace").strip()
            if not resp_body:
                passthrough()
            try:
                result = json.loads(resp_body)
                print(json.dumps(result))
                sys.exit(0)
            except json.JSONDecodeError:
                passthrough()
    except (urllib.error.URLError, urllib.error.HTTPError, OSError, TimeoutError):
        if fail_closed:
            deny("[http-hook] Enterprise DLP server unreachable and CCA_HTTP_HOOK_FAIL_CLOSED=1. Blocking action.")
        passthrough()


if __name__ == "__main__":
    main()
